airflow/contrib/hooks/sftp_hook.py

```diff
@@ -0,0 +1,141 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import pysftp
+import logging
+import datetime
+from airflow.hooks.base_hook import BaseHook
+
+class SFTPHook(BaseHook):
+    """
+    Interact with SFTP. Aims to be interchangeable with FTPHook.
+
+    Pitfalls: - In contrast with FTPHook describe_directory only returns size and modify. It doesn't return unix.owner,
+                unix.mode, perm, unix.group, unique and type.
+              - retrieve_file and store_file only take a local full path and not a buffer.
+              - If no mode is passed to create_directory it will be created with 777 permissions.
+
+    Errors that may occur throughout but should be handled
+    downstream.
+    """
+
+    def __init__(self, ftp_conn_id):
+        self.ftp_conn_id = ftp_conn_id
+        self.conn = None
+
+    def get_conn(self):
+        """
+        Returns an SFTP connection object
+        """
+        if self.conn is None:
+            params = self.get_connection(self.ftp_conn_id)
+            self.conn = pysftp.Connection(host=params.host, port=params.port, username=params.login, password=params.password)
+
+        return self.conn
+
+    def close_conn(self):
+        """
+        Closes the connection. An error will occur if the
+        connection wasnt ever opened.
+        """
+        conn = self.conn
+        conn.close()
```

Comment:
Should you set `self.conn = None` after this?  I would have expected the following block of code to work:

```
hook = SFTPHook(conn_id)
hook.list_directory('/')
hook.close()
hook.get_conn() # note: this is implicit if you had just called the next line
hook.list_directory('/')
```